src/cpp/desktop/DesktopMain.cpp

```diff
@@ -827,6 +828,39 @@ int main(int argc, char* argv[])
       if (!prepareEnvironment(options))
          return 1;
 
+#ifdef __APPLE__
+      // on macOS, we need to figure out whether we're about to run
+      // with an x86_64 or an arm64 build of R, and select the appropriate
+      // rsession copy based on that
+      {
+         FilePath rHome = FilePath(core::system::getenv("R_HOME"));
+         FilePath rLib = rHome.completeChildPath("lib/libR.dylib");
+
+         core::system::ProcessOptions options;
+         core::system::ProcessResult result;
+         Error lipoError = core::system::runProgram(
+                  "/usr/bin/lipo",
+                  { "-archs", rLib.getAbsolutePath() },
+                  std::string(),
+                  options,
+                  &result);
+
+         if (error)
+         {
+            LOG_ERROR(error);
+         }
+
+         std::string rArchs = string_utils::trimWhitespace(result.stdOut);
+         if (rArchs.find("arm64") != std::string::npos)
+         {
+            FilePath arm64SessionPath = sessionPath.getParent().completeChildPath("rsession-arm64");
+            if (arm64SessionPath.exists())
+               sessionPath = arm64SessionPath;
```

Comment:
We should log a warning or error here as there is almost certainly trouble ahead.